orm/model/ormtable/bench_test.go

```diff
@@ -0,0 +1,315 @@
+	"gotest.tools/v3/assert"
+
+	"github.com/cosmos/cosmos-sdk/orm/internal/testpb"
+	"github.com/cosmos/cosmos-sdk/orm/model/ormtable"
+	"github.com/cosmos/cosmos-sdk/orm/types/kv"
+)
+
+var balanceTable testpb.BalanceTable
+
+func init() {
+	table, err := ormtable.Build(ormtable.Options{
+		MessageType: (&testpb.Balance{}).ProtoReflect().Type(),
+	})
+	if err != nil {
+		panic(err)
+	}
+
+	balanceTable, err = testpb.NewBalanceTable(table)
+	if err != nil {
+		panic(err)
+	}
+}
+
+func BenchmarkMemory(b *testing.B) {
+	bench(b, func(tb testing.TB) ormtable.Backend {
+		return ormtest.NewMemoryBackend()
+	})
+}
+
+func BenchmarkLevelDB(b *testing.B) {
+	bench(b, testkv.NewGoLevelDBBackend)
+}
+
+func bench(b *testing.B, newBackend func(testing.TB) ormtable.Backend) {
+	b.Run("insert", func(b *testing.B) {
+		b.StopTimer()
+		ctx := ormtable.WrapContextDefault(newBackend(b))
+		b.StartTimer()
+		benchInsert(b, ctx)
+	})
+	b.Run("update", func(b *testing.B) {
+		b.StopTimer()
+		ctx := ormtable.WrapContextDefault(newBackend(b))
+		benchInsert(b, ctx)
+		b.StartTimer()
+		benchUpdate(b, ctx)
+	})
+	b.Run("get", func(b *testing.B) {
+		b.StopTimer()
+		ctx := ormtable.WrapContextDefault(newBackend(b))
+		benchInsert(b, ctx)
+		b.StartTimer()
+		benchGet(b, ctx)
+	})
+	b.Run("delete", func(b *testing.B) {
+		b.StopTimer()
+		ctx := ormtable.WrapContextDefault(newBackend(b))
+		benchInsert(b, ctx)
+		b.StartTimer()
+		benchDelete(b, ctx)
+	})
+}
+
+func benchInsert(b *testing.B, ctx context.Context) {
+	for i := 0; i < b.N; i++ {
+		assert.NilError(b, balanceTable.Insert(ctx, &testpb.Balance{
+			Address: fmt.Sprintf("acct%d", i),
+			Denom:   "bar",
+			Amount:  10,
+		}))
+	}
+}
+
+func benchUpdate(b *testing.B, ctx context.Context) {
+	for i := 0; i < b.N; i++ {
+		assert.NilError(b, balanceTable.Update(ctx, &testpb.Balance{
+			Address: fmt.Sprintf("acct%d", i),
+			Denom:   "bar",
+			Amount:  11,
+		}))
+	}
+}
+
+func benchGet(b *testing.B, ctx context.Context) {
+	for i := 0; i < b.N; i++ {
+		balance, err := balanceTable.Get(ctx, fmt.Sprintf("acct%d", i), "bar")
+		assert.NilError(b, err)
+		assert.Equal(b, uint64(10), balance.Amount)
+	}
+}
+
+func benchDelete(b *testing.B, ctx context.Context) {
+	for i := 0; i < b.N; i++ {
+		assert.NilError(b, balanceTable.Delete(ctx, &testpb.Balance{
+			Address: fmt.Sprintf("acct%d", i),
+			Denom:   "bar",
+		}))
+	}
+}
+
+//
+// Manually written versions of insert, update, delete and get for testpb.Balance
+//
+
+const (
+	addressDenomPrefix byte = iota
+	denomAddressPrefix
+)
+
+func insertBalance(store kv.Store, balance *testpb.Balance) error {
+	denom := balance.Denom
+	balance.Denom = ""
+	addr := balance.Address
+	balance.Address = ""
+
+	addressDenomKey := []byte{addressDenomPrefix}
+	addressDenomKey = append(addressDenomKey, []byte(addr)...)
+	addressDenomKey = append(addressDenomKey, 0x0)
+	addressDenomKey = append(addressDenomKey, []byte(denom)...)
+	has, err := store.Has(addressDenomKey)
```

Comment:
Note: ideally a store would provide insert with check function - in `store/v2` this is a nice and easy optimization.

Reply:
How about a `Set` method which returns `existingValue []byte`?